clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java

```diff
@@ -376,25 +376,21 @@ protected Map<Node, FetchSessionHandler.FetchRequestData> prepareCloseFetchSessi
         final Cluster cluster = metadata.fetch();
         Map<Node, FetchSessionHandler.Builder> fetchable = new HashMap<>();
 
-        try {
-            sessionHandlers.forEach((fetchTargetNodeId, sessionHandler) -> {
-                // set the session handler to notify close. This will set the next metadata request to send close message.
-                sessionHandler.notifyClose();
+        sessionHandlers.forEach((fetchTargetNodeId, sessionHandler) -> {
```

Comment:
seems like you took out the `sessionHandlers.clear();` in the finally block - is there any reason why we don't need it?

Reply:
Yes, if we clear the sessions, we potentially end up with inflight response handlers that fail to find their corresponding session handler. This lookup failure results in the error log line reported in the Jira.

Reply:
There are some potential alternatives to solve this problem, but the path I took was to simply revert to the previous behavior which did _not_ clear the cache.

Reply:
I think that's worth exploring, because I don't see that we ever remove any of the `FetchSessionHandler` entries once they're created 😬

Reply:
We don't remove the `sessionHandlers` map explicitly, no. It's garbage collected when the `AbstractFetch`/`Fetcher` object itself is garbage collected.

Reply:
@philipnee—it doesn't look like there's actually an issue with respect to stale partition data in the `FetchSessionHandler`. It's reset on every call to [`prepareFetchRequests()`](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L406).

The `FetchSessionHandler` uses an inner `Builder` class to keep track of the partitions the user is requesting. For each distinct fetch cycle [a new `Builder` is created](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L443). Each partition that is requested is [added to the newly-created `Builder` instance](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L452). When all the [fetchable partitions](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L324) have been added, [the `Builder.build()` method is called](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L459) which creates the `FetchRequestData`.

So the data kept in the `FetchSessionHandler` is ephemeral. Any partition leadership changes would happen somewhere upstream and get updated in the `SubscriptionState`, well before we get to the fetch request generation.

Reply:
> There are some potential alternatives to solve this problem, but the path I took was to simply revert to the previous behavior which did not clear the cache.

I think that the cache was actually cleared in `Fetcher.close()` before the refactoring. I suppose that we could bring it back too even if it does not bring much in the end.